tests/unit/hooks/entrypoint_hook.py

```diff
@@ -0,0 +1,169 @@
+        - self.reference store the expected Command for comparison.
+
+        Stored Command objects are comparable, used for asserts.
+        Example:
+        >>> assert hook.result == hook.reference
+        """
+        # Clean hook from previous test, store the command to test
+        self._reset_attributes()
+
+        # Default environment variables needed by all tests
+        test_environ.update(self.DEFAULT_ENV)
+        result_environ.update(self.DEFAULT_ENV)
+
+        # Manage system arguments & environment used by the script
+        sys.argv[1:] = test_argv.copy()
+        os.environ = test_environ.copy()
+
+        # Run the test, launching entrypoint script from the main
+        entrypoint.main()
+
+        # Store expected Command used for comparison
+        self.reference = Command(result_argv, result_environ)
+
+        if self.result is None:
+            raise CommandNotFound("Test fail, do not return a command")
+
+    def _execve_hook(self, executable, argv, environ):
+        """
+        Hook for os.execve function, to catch arguments/environment
+        instead of launching processes.
+        """
+        assert executable == argv[0]
+        self.result = Command(argv, environ)
+
+    @staticmethod
+    def _get_help_hook(command_arguments):
+        """
+        Hook call of executable help menu to retrieve options.
+        Fake a list of raw options generated by entrypoint.get_help.
+        """
+        executable = command_arguments[0]
+
+        #Test use of -help-debug to expand help options
+        if executable == "dogecoind":
+            assert "-help-debug" in command_arguments
+        else:
+            assert "-help-debug" not in command_arguments
+        return getattr(hooks.help_menus, executable.replace("-", "_"))
+
+    def _reset_attributes(self):
+        """Clean state between each test"""
+        self.result = None
+        self.reference = None
+
+    def _setup_hooks(self):
+        """
+        Enable hooks of entrypoint.py system & functions calls, disable
+        some functions.
+
+        Replace entrypoint function by EntrypointHook methods
+        to handle arguments used by entrypoint calls.
+
+        Save references to previous functions to restore them test
+        clean up.
+        """
+        # Save hooked functions for restoration
+        self._execve_backup = os.execve
+        self._setgid_backup = os.setgid
+        self._setuid_backup = os.setuid
+
+        # Save function hooked only on the host
+        self._which_backup = shutil.which
+        self._get_help_backup = entrypoint.get_help
+
+        # Setup hooks
+        # Add execve hook globally to catch entrypoint arguments
+        os.execve = self._execve_hook
+
+        # Disable setgid & setuid behavior
+        os.setgid = lambda _ : None
+        os.setuid = lambda _ : None
+
+        # Extra hooks to run tests from host, replace executables availability
+        if self.CONTAINER_MODE is False:
```

Comment:
I really think this should not be conditional. It may look fancy, but it's better if unit tests are agnostic to the environment they're ran in - we should focus on testing our code here, not testing whether `shutil.which` works or if `dogecoind` returns a usable help menu. This we can test in the integration tests.